src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2573,17 +2495,21 @@ private bool Fix(int iParam, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
             return _getIsNullableOpt?.Invoke(expr);
         }
 
-        internal static TypeSymbol Merge(TypeSymbol first, TypeSymbol second, AssemblySymbol corLibrary)
+        internal static TypeSymbolWithAnnotations Merge(TypeSymbolWithAnnotations first, TypeSymbolWithAnnotations second, VarianceKind variance, ConversionsBase conversions, out bool hadNullabilityMismatch)
         {
-            var firstWithAnnotations = TypeSymbolWithAnnotations.Create(first);
-            var secondWithAnnotations = TypeSymbolWithAnnotations.Create(second);
-            return MergeNullability(MergeTupleNames(MergeDynamic(firstWithAnnotations, secondWithAnnotations, corLibrary), secondWithAnnotations), secondWithAnnotations).TypeSymbol;
+            var merged = MergeTupleNames(MergeDynamic(first, second, conversions.CorLibrary), second);
+            if (!conversions.IncludeNullability)
+            {
+                hadNullabilityMismatch = false;
+                return merged.SetUnknownNullabilityForReferenceTypes();
+            }
```

Comment:
I think we want to preserve the nullability computed during initial binding if it was not a result of type inference. To do that we will need a fourth nullability value - `NotComputed` - as part of the computation.  In which case we would have to stop using `bool?` to represent nullability throughout the compiler.  That is too big a change for this PR. So please just note that this is one place where the change will be nontrivial.

Reply:
Logged [https://github.com/dotnet/roslyn/issues/30534](https://github.com/dotnet/roslyn/issues/30534).

---
In reply to: [225360955](https://github.com/dotnet/roslyn/pull/29727#discussion_r225360955) [](ancestors = 225360955)